auth0/src/test/java/com/auth0/android/dpop/DPoPProviderTest.kt

```diff
@@ -0,0 +1,421 @@
+
+        val result =
+            DPoPProvider.getHeaderData(testHttpMethod, testHttpUrl, testAccessToken, tokenType)
+
+        assertThat(result.authorizationHeader, `is`("DPoP $testAccessToken"))
+        assertThat(result.dpopProof, `is`(notNullValue()))
+        val proof = Jwt(result.dpopProof!!)
+
+        Assert.assertEquals(proof.decodedHeader["typ"] as String, algTyp)
+        Assert.assertEquals(proof.decodedHeader["alg"] as String, alg)
+        Assert.assertEquals(
+            (proof.decodedHeader["jwk"] as LinkedTreeMap<*, *>).toString(),
+            testProofJwk
+        )
+        Assert.assertEquals(proof.decodedPayload["htm"] as String, testHttpMethod)
+        Assert.assertEquals(proof.decodedPayload["htu"] as String, testHttpUrl)
+        Assert.assertEquals(proof.decodedPayload["ath"] as String, testEncodedAccessToken)
+    }
+
+    @Test
+    public fun `getHeaderData should return DPoP token with proof including nonce when provided`() {
+        val tokenType = "DPoP"
+        whenever(mockKeyStore.hasKeyPair()).thenReturn(true)
+        whenever(mockKeyStore.getKeyPair()).thenReturn(Pair(fakePrivateKey, fakePublicKey))
+
+        val result = DPoPProvider.getHeaderData(
+            testHttpMethod,
+            testHttpUrl,
+            testAccessToken,
+            tokenType,
+            testNonce
+        )
+
+        assertThat(result.authorizationHeader, `is`("DPoP $testAccessToken"))
+        assertThat(result.dpopProof, `is`(notNullValue()))
+
+        val proof = Jwt(result.dpopProof!!)
+        Assert.assertEquals(proof.decodedHeader["typ"] as String, algTyp)
+        Assert.assertEquals(proof.decodedHeader["alg"] as String, alg)
+        Assert.assertEquals(
+            (proof.decodedHeader["jwk"] as LinkedTreeMap<*, *>).toString(),
+            testProofJwk
+        )
+        Assert.assertEquals(proof.decodedPayload["htm"] as String, testHttpMethod)
+        Assert.assertEquals(proof.decodedPayload["htu"] as String, testHttpUrl)
+        Assert.assertEquals(proof.decodedPayload["ath"] as String, testEncodedAccessToken)
+        Assert.assertEquals(proof.decodedPayload["nonce"] as String, testNonce)
+    }
+
+    @Test
+    public fun `isNonceRequiredError should return true for 400 response with nonce required error`() {
+        whenever(mockResponse.peekBody(Long.MAX_VALUE)).thenReturn("{\"error\":\"use_dpop_nonce\"}".toResponseBody())
+        whenever(mockResponse.code).thenReturn(400)
+
+        val result = DPoPProvider.isNonceRequiredError(mockResponse)
+        assertThat(result, `is`(true))
+    }
+
+    @Test
+    public fun `isNonceRequiredError should return true for 401 response with resource server nonce error`() {
+        whenever(mockResponse.code).thenReturn(401)
+        whenever(mockResponse.headers).thenReturn(
+            Headers.Builder().add("WWW-Authenticate", "DPoP error=\"use_dpop_nonce\"").build()
```

Comment:
There should be another test case for when `use_dpop_nonce` is not quoted.

Reply:
Also, another test case with a `WWW-Authenticate` header containing both `Bearer` and `DPoP` challenges.